lib/IRGen/HeapTypeInfo.h

```diff
@@ -117,59 +122,85 @@ class HeapTypeInfo : public SingleScalarTypeInfo<Impl, ReferenceTypeInfo> {
-    llvm::Value *value = e.claimNext();
-    IGF.emitStrongRetainUnowned(value, asDerived().getReferenceCounting(),
-                                atomicity);
-  }
-
-  void strongRetainUnownedRelease(IRGenFunction &IGF,
-                                  Explosion &e,
-                                  Atomicity atomicity) const override {
-    llvm::Value *value = e.claimNext();
-    IGF.emitStrongRetainAndUnownedRelease(value,
-                                          asDerived().getReferenceCounting(),
-                                          atomicity);
-  }
-
-  void unownedRetain(IRGenFunction &IGF, Explosion &e,
-                     Atomicity atomicity) const override {
-    llvm::Value *value = e.claimNext();
-    IGF.emitUnownedRetain(value, asDerived().getReferenceCounting(), atomicity);
-  }
-
-  void unownedRelease(IRGenFunction &IGF, Explosion &e,
-                      Atomicity atomicity) const override {
-    llvm::Value *value = e.claimNext();
-    IGF.emitUnownedRelease(value, asDerived().getReferenceCounting(), atomicity);
-  }
-
-  void unownedLoadStrong(IRGenFunction &IGF, Address src,
-                         Explosion &out) const override {
-    llvm::Value *value = IGF.emitUnownedLoadStrong(src, this->getStorageType(),
-                                           asDerived().getReferenceCounting());
-    out.add(value);
-  }
-
-  void unownedTakeStrong(IRGenFunction &IGF, Address src,
-                         Explosion &out) const override {
-    llvm::Value *value = IGF.emitUnownedTakeStrong(src, this->getStorageType(),
-                                           asDerived().getReferenceCounting());
-    out.add(value);
-  }
-
-  void unownedInit(IRGenFunction &IGF, Explosion &in,
-                   Address dest) const override {
-    IGF.emitUnownedInit(in.claimNext(), dest,
-                        asDerived().getReferenceCounting());
-  }
-
-  void unownedAssign(IRGenFunction &IGF, Explosion &in,
-                     Address dest) const override {
-    IGF.emitUnownedAssign(in.claimNext(), dest,
-                          asDerived().getReferenceCounting());
-  }
+#define ADDRESS_ONLY_DYNAMIC_REF_STORAGE(Name, name, ...) \
+  const TypeInfo * \
+  create##Name##StorageType(TypeConverter &TC) const override { \
+    return TC.create##Name##StorageType(this->getStorageType(), \
+                                        asDerived().getReferenceCounting()); \
+  } \
+  void name##LoadStrong(IRGenFunction &IGF, Address src, \
+                         Explosion &out) const override { \
+    llvm::Value *value = IGF.emit##Name##LoadStrong(src, \
+                                          this->getStorageType(), \
+                                          asDerived().getReferenceCounting()); \
+    if (optionalityOf(ReferenceOwnership::Name) == Optionality::Required) { \
+      out.add(IGF.Builder.CreatePtrToInt(value, getOptionalIntType())); \
+    } else { \
+      out.add(value); \
+    } \
```

Comment:
This should be based on the actual optionality of the referent type, not whether the reference kind generally supports optionality.  Otherwise it's not going to do the right thing for reference types whose referent type can be either optional or not.

Reply:
Thanks for clarifying. I wasn't sure about this and everything seemed to work with the code at hand.

How can one dig out the optionality state based on the data structures at hand? Is it doable? Or do I need to push an "is optional" boolean through the design?

Reply:
It probably wouldn't be the worst thing to pass a SILType to all the reference-counting operations, or yeah, you could store a boolean when creating it.